ctap2token/token.go

```diff
@@ -0,0 +1,663 @@
+	0x02: ErrInvalidParameter,
+	0x03: ErrInvalidLength,
+	0x04: ErrInvalidSeq,
+	0x05: ErrTimeout,
+	0x06: ErrChannelBusy,
+	0x0A: ErrLockRequired,
+	0x0B: ErrInvalidChannel,
+	0x11: ErrCborUnexpectedType,
+	0x12: ErrInvalidCbor,
+	0x14: ErrMissingParameter,
+	0x15: ErrLimitExceeded,
+	0x16: ErrUnsupportedExtension,
+	0x19: ErrCredentialExcluded,
+	0x21: ErrProcessing,
+	0x22: ErrInvalidCredential,
+	0x23: ErrUserActionPending,
+	0x24: ErrOperationPending,
+	0x25: ErrNoOperations,
+	0x26: ErrUnsupportedAlgorithm,
+	0x27: ErrOperationDenied,
+	0x28: ErrKeyStoreFull,
+	0x2A: ErrNoOperationPending,
+	0x2B: ErrUnsupportedOption,
+	0x2C: ErrInvalidOption,
+	0x2D: ErrKeepaliveCancel,
+	0x2E: ErrNoCredentials,
+	0x2F: ErrUserActionTimeout,
+	0x30: ErrNotAllowed,
+	0x31: ErrPinInvalid,
+	0x32: ErrPinBlocked,
+	0x33: ErrPinAuthInvalid,
+	0x34: ErrPinAuthBlocked,
+	0x35: ErrPinNotSet,
+	0x36: ErrPinRequired,
+	0x37: ErrPinPolicyViolation,
+	0x38: ErrPinTokenExpired,
+	0x39: ErrRequestTooLarge,
+	0x3A: ErrActionTimeout,
+	0x3B: ErrUpRequired,
+	0xDF: ErrSpecLast,
+	0xE0: ErrExtensionFirst,
+	0xEF: ErrExtensionLast,
+	0xF0: ErrVendorFirst,
+	0xFF: ErrVendorLast,
+}
+
+type Device interface {
+	// CBOR sends a CTAP2 CBOR encoded message to the device and returns the response.
+	CBOR(data []byte) ([]byte, error)
+	// Message sends a CTAP1 message to the device and returns the response.
+	Message(data []byte) ([]byte, error)
+	// SetResponseTimeout sets the maximum time to wait for a response from the device.
+	SetResponseTimeout(timeout time.Duration)
+	Cancel()
+	Close()
+}
+
+// NewToken returns a token that will use Device to communicate with the device.
+func NewToken(d Device) *Token {
+	return &Token{d: d}
+}
+
+// A Token implements the FIDO U2F hardware token messages as defined in the Raw
+// Message Formats specification.
+type Token struct {
+	d Device
+}
+
+type MakeCredentialRequest struct {
+	ClientDataHash   ClientDataHash          `cbor:"1,keyasint"`
+	RP               CredentialRpEntity      `cbor:"2,keyasint"`
+	User             CredentialUserEntity    `cbor:"3,keyasint"`
+	PubKeyCredParams []CredentialParam       `cbor:"4,keyasint"`
+	ExcludeList      []CredentialDescriptor  `cbor:"5,keyasint,omitempty"`
+	Extensions       AuthenticatorExtensions `cbor:"6,keyasint,omitempty"`
+	Options          AuthenticatorOptions    `cbor:"7,keyasint,omitempty"`
+	// PinUVAuth is the first 16 bytes of HMAC-SHA-256 of clientDataHash using
+	// pinToken which platform got from the authenticator
+	// we need a pointer here to distinguish nil pinUVAuth from empty.
+	// When nil, it must be omitted from the CBOR encoded request, but included when empty,
+	PinUVAuth *[]byte `cbor:"8,keyasint,omitempty"`
+	// PinUVAuthProtocol is the PIN protocol version chosen by the client
+	PinUVAuthProtocol PinUVAuthProtocolVersion `cbor:"9,keyasint,omitempty"`
+}
+
+// MakeCredentialResponse
+// TODO: structure may be different with different kind of attestations.
```

Comment:
Hmm no. The MakeCredentialResponse structure is fixed and only the content of AttSmt may vary, but since it's a map[string]interface{} we should be good to store anything. Thanks, removed.